aidb/engine/approx_select_engine.py

```diff
@@ -0,0 +1,225 @@
+
+    estimated_z1 = list(z[:positive_length]) + [0] * (total_length - len(z[:positive_length]))
+    estimated_z2 = list(z[positive_length:]) + [0] * (total_length - len(z[positive_length:]))
+    z1_mean, z1_std = np.mean(estimated_z1), np.std(estimated_z1)
+    z2_mean, z2_std = np.mean(estimated_z2), np.std(estimated_z2)
+
+    delta = 1.0 - confidence
+
+    _, ub = self._get_confidence_bounds(z1_mean, z1_std, len(estimated_z1), delta / 2)
+    lb, _ = self._get_confidence_bounds(z2_mean, z2_std, len(estimated_z2), delta / 2)
+
+    # inflate recall to correct the sampling errors
+    if (ub+lb) == 0.0:
+      modified_recall_target = 1.0
+    else:
+      modified_recall_target = ub / (ub + lb)
+
+    modified_tau = self.tau_estimate(
+      modified_recall_target * sum(z),
+      satisfied_sampled_results
+    )
+
+    logger.info(f'modified_tau: {modified_tau}')
+    return modified_tau
+
+
+  async def execute_approx_select_query(self, query: Query):
+    if not query.is_valid_approx_select_query:
+      raise Exception('Approx select query should contain Confidence and should not contain Budget.')
+
+    # generate proxy score for each blob
+    score_for_all_df, score_connected = await self.get_proxy_scores_for_all_blobs(query)
+
+    proxy_score_for_all_blobs = self.score_fn(score_for_all_df, score_connected)
+
+    dataset = self.get_sampled_proxy_blob(proxy_score_for_all_blobs)
+
+    # This is used for parallel test
+    seed = (mp.current_process().pid * np.random.randint(100000, size=1)[0]) % (2**32 - 1)
+    sampled_df = dataset.sample(BUDGET, replace=True, weights=WEIGHT, random_state=seed)
+
+    async with self._sql_engine.begin() as conn:
+      satisfied_sampled_results, all_sampled_results = await self.get_inference_results(
+          query,
+          list(sampled_df.index),
+          conn
+      )
+
+      satisfied_sampled_results = satisfied_sampled_results.join(sampled_df, how='inner')
+      sorted_satisfied_sampled_results = satisfied_sampled_results.sort_values(by=PROXY_SCORE, ascending=False)
+
+      no_result_length = BUDGET - len(sampled_df.loc[list(set(all_sampled_results.index))])
+
+      all_sampled_results = all_sampled_results.join(sampled_df, how='inner')
+
+      total_length = no_result_length + len(all_sampled_results)
+
+      recall_target = query.recall_target
+      confidence = query.confidence / 100
+      tau_modified = self.tau_modified(
+          sorted_satisfied_sampled_results,
+          recall_target,
+          confidence,
+          total_length
+      )
+
+      R1 = sorted_satisfied_sampled_results.index
+      R2 = dataset[dataset[PROXY_SCORE] >= tau_modified].index
+
+      additional_samples = list(set(R1).union(set(R2)))
+
+      logger.info(f'num_samples: {len(additional_samples)}')
+
+      additional_satisfied_sampled_results,  additional_all_sampled_results = await self.get_inference_results(
```

Comment:
Don't we also want to take the positive records from the pilot sample?

Reply:
Yes, we have taken the positive records from the pilot sample. The variable names are ambiguous, I will change them.
https://github.com/ddkang/aidb/blob/4c677999b78afae88541f01d2d80db6d26bd5877/aidb/engine/approx_select_engine.py#L212-L215